libs/cardano-ledger-core/src/Cardano/Ledger/State/StakePool.hs

```
@@ -0,0 +1,467 @@
+      , spsMargin = def
+      , spsRewardAccount = def
+      , spsOwners = def
+      , spsRelays = def
+      , spsMetadata = def
+      }
+
+-- | Convert 'PoolParams' to 'StakePoolState' by dropping the pool ID.
+-- This is the primary way to create a 'StakePoolState' from registration
+-- or update parameters.
+mkStakePoolState :: PoolParams -> StakePoolState
+mkStakePoolState pp =
+  StakePoolState
+    { spsVrf = ppVrf pp
+    , spsPledge = ppPledge pp
+    , spsCost = ppCost pp
+    , spsMargin = ppMargin pp
+    , spsRewardAccount = ppRewardAccount pp
+    , spsOwners = ppOwners pp
+    , spsRelays = ppRelays pp
+    , spsMetadata = ppMetadata pp
+    }
+
+-- | Convert 'StakePoolState' back to 'PoolParams' by providing the pool ID.
+-- This is useful when you need to reconstruct the full parameters from
+-- the state representation.
+stakePoolStateToPoolParams :: KeyHash 'StakePool -> StakePoolState -> PoolParams
+stakePoolStateToPoolParams poolId sps =
+  PoolParams
+    { ppId = poolId
+    , ppVrf = spsVrf sps
+    , ppPledge = spsPledge sps
+    , ppCost = spsCost sps
+    , ppMargin = spsMargin sps
+    , ppRewardAccount = spsRewardAccount sps
+    , ppOwners = spsOwners sps
+    , ppRelays = spsRelays sps
+    , ppMetadata = spsMetadata sps
+    }
+
+data PoolMetadata = PoolMetadata
+  { pmUrl :: !Url
+  , pmHash :: !ByteString
+  }
+  deriving (Eq, Ord, Generic, Show)
+
+deriving instance NFData PoolMetadata
+
+instance ToJSON PoolMetadata where
+  toJSON pmd =
+    Aeson.object
+      [ "url" .= pmUrl pmd
+      , "hash" .= Text.decodeLatin1 (B16.encode (pmHash pmd))
+      ]
+
+instance FromJSON PoolMetadata where
+  parseJSON =
+    Aeson.withObject "PoolMetadata" $ \obj -> do
+      url <- obj .: "url"
+      hash <- explicitParseField parseJsonBase16 obj "hash"
+      return $ PoolMetadata url hash
+
+parseJsonBase16 :: Value -> Parser ByteString
+parseJsonBase16 v = do
+  s <- parseJSON v
+  case B16.decode (Char8.pack s) of
```

There is actually a bug in here that would silently convert unicode to ascii, potentially causing invalid symbols becoming valid. Let's fix it:

```suggestion
  txt <- parseJSON v
  unless (Text.isAscii txt) $ fail $ "Supplied text contains non-ASCII characters: " <> show txt
  case B16.decode (Text.encodeUtf8 txt) of
```
Could you please fix it in a separate commit.